### DISCUSSION TITLE
[App Router] How to share data across pages / components on the server side?

### DISCUSSION THREAD

**Discussion by weilueluo** [vercel/next.js#50729] [Help] [answered]
### Summary

I have a JSON file to be fetched on the server side that will be used across all my pages, I tried to do this using a react context provider in the root layout, but it throws an error saying that `createContext` only works in client components, so what is the right way to do this?

**Comment by devjiwonchoi** [comment]
When you have an `error` with `createContext` only works in client components, it is because your component is Server Component but trying to use the client-only feature.

In this case, try to add `'use client';` on the top of the file.

> The "use client" directive must be defined at the top of a file before any imports.

Ref: https://nextjs.org/docs/getting-started/react-essentials#client-components

Although this might fix the problem, is there a reason why you are not using the [server-side fetch](https://nextjs.org/docs/app/building-your-application/data-fetching#fetching-data-on-the-server)?

**Reply by weilueluo** [reply]
Hi @devjiwonchoi , thanks for the reply. Sorry for my ignorance but if I add `'use client'` at the top, won't it cause the context to be used only on the client? I want to share the JSON file during server-side rendering.

**Reply by devjiwonchoi** [reply]
Right, and that was my doubt.

If you want to fetch data server-side, why aren't you using the server-side fetch as stated before?

**Reply by weilueluo** [reply]
hmm, I do not see why I need to use server-side fetch, because the file exists locally, I am doing a `fs.readFileSync` to retrieve it. Do you mean I should put this code in the `/api/route.js` and use `fetch` to retrieve the data when I needed it?

**Reply by weilueluo** [reply, marked as answer]
Sorry, I think I did not phrase my problem properly. This is a duplicate of https://github.com/vercel/next.js/discussions/45543
